feat(tools, client): add speaking tasks logic

DanielRosa74 opened the pull request:
Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [X] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [X] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/how-to-open-a-pull-request/).
- [X] My pull request targets the `main` branch of freeCodeCamp.
- [X] I have tested these changes either locally on my machine, or Gitpod.

This PR adds an optional (in the sense that you don't need to get the answer right or even do its exercise to proceed to the next task) speaking modal to some tasks from the English curriculum. At the moment, the only task you will be able to test this locally is /learn/b1-english-for-developers/learn-about-adverbial-phrases/task-19. The other speaking tasks still need their respective audio files to be uploaded to their repository.

<details>
<summary>Screenshots</summary>

<img width="805" height="272" alt="Screenshot 2025-09-10 at 17 31 47" src="https://github.com/user-attachments/assets/466672fc-2f66-43f9-80c7-65edacce25a6" />

<img width="966" height="683" alt="Screenshot 2025-09-10 at 17 31 58" src="https://github.com/user-attachments/assets/eef0663f-4885-4ced-b48d-b04930f19f05" />

<img width="938" height="429" alt="Screenshot 2025-09-10 at 17 32 23" src="https://github.com/user-attachments/assets/029a7044-732c-4d00-aa6c-cd6e18d2a36a" />

<img width="926" height="428" alt="Screenshot 2025-09-10 at 17 32 59" src="https://github.com/user-attachments/assets/2ba4bf12-7376-4094-8589-779bd959c1c6" />

</details>

DanielRosa74 committed:
add button and modal for speaking tasks

DanielRosa74 committed:
Fix speakiing button style

DanielRosa74 committed:
add buttons and feedback area to modal

DanielRosa74 committed:
Make speaking modal component

DanielRosa74 committed:
Improve logic changes

DanielRosa74 committed:
Implement the recording and playback logic

DanielRosa74 committed:
Remove debug from show.tsx

DanielRosa74 committed:
add last-minute styling changes

DanielRosa74 committed:
fix lint errors

DanielRosa74 commented on client/src/templates/Challenges/components/speaking-modal.css:
```diff
@@ -0,0 +1,102 @@
+  text-align: center;
+  color: #ffd700;
+  font-size: 1.1rem;
+}
```
```suggestion
}

```

DanielRosa74 committed:
fix more linting issues

DanielRosa74 commented on client/src/templates/Challenges/generic/show.tsx:
```diff
@@ -272,12 +272,18 @@ const ShowGeneric = ({
-                    submittedMcqAnswers={submittedMcqAnswers}
-                    showFeedback={showFeedback}
-                  />
+                  questions={questions}
+                  selectedOptions={selectedMcqOptions}
+                  handleOptionChange={handleMcqOptionChange}
+                  submittedMcqAnswers={submittedMcqAnswers}
+                  showFeedback={showFeedback}
+                  showSpeakingButton={showSpeakingButton}
+                  challengeData={{
+                    audio: {
+                      filename: scene?.setup?.audio?.filename
+                    }
+                  }}
+                />
```
```suggestion
                    questions={questions}
                    selectedOptions={selectedMcqOptions}
                    handleOptionChange={handleMcqOptionChange}
                    submittedMcqAnswers={submittedMcqAnswers}
                    showFeedback={showFeedback}
                    showSpeakingButton={showSpeakingButton}
                    challengeData={{
                      audio: {
                        filename: scene?.setup?.audio?.filename
                      }
                    }}
                  />
```

DanielRosa74 committed:
add finding audio file logic at the cdn

DanielRosa74 commented on client/src/templates/Challenges/generic/show.tsx:
```diff
@@ -10,6 +10,7 @@ import { YouTubeEvent } from 'react-youtube';
 import { ObserveKeys } from 'react-hotkeys';
```
```suggestion
```

DanielRosa74 replied:
```suggestion
```

DanielRosa74 committed:
fix get-challenges attributing showSpeakingButton to all challenges

ojeytonwilliams reviewed (changes requested):
Hey Daniel, I've been playing around with this a little and noticed a few strange bugs: 

* If you press record and immediately stop it gives you the feedback "Recording stopped. Processing..." and never updates. 

* If you press record and say nothing, you get the error `Speech recognition error: no-speech` rather than `Recording stopped. No speech detected.`.

* If you make some noise you either get `Recording stopped. No speech detected.` (i.e. what you want) or stuck on `Recording... Speak now.`, even after you stop making noise.

Other than that I think it works well. I do have some feedback about the structure, though:

DanielRosa74 committed:
implement feedback suggestions

DanielRosa74 committed:
fix tests

DanielRosa74 committed:
change finding audiofile logic

DanielRosa74 committed:
fix recording issues

ojeytonwilliams commented on client/src/templates/Challenges/components/speaking-modal.tsx:
```diff
@@ -0,0 +1,439 @@
+    } catch (error) {
+      console.error('Error setting up audio monitoring:', error);
+    }
+  };
```
Hey, this conversation was resolved prematurely. Also, while I'm here: I think analyser.fftSize = 256; deserves a comment since it's not clear why `256` was chosen.

DanielRosa74 committed:
Add Oliver's implementations

DanielRosa74 committed:
audioId changes

DanielRosa74 committed:
Add fix for last commit not passing tests

DanielRosa74 committed:
fix failing test

DanielRosa74 committed:
Fix to get-challenges.js

DanielRosa74 committed:
add showSpeakingButton to challenge-schema.js

DanielRosa74 committed:
add audioId to challenge-schema.js

huyenltnguyen reviewed (commented):
I didn’t leave comments on every line, but here are some general suggestions:
- We should try to use components from `freecodecamp/ui` where possible.
- Styles should be moved to the CSS file instead of using inline styles.
- Strings should be added to `translations.json` for localization.

Let me know if you’d like any clarification or if you'd like me to help with the updates :)

DanielRosa74 committed:
Style changes implemented as per Huyen's feedback

DanielRosa74 committed:
Set buttons style to default

DanielRosa74 committed:
Transform speaking-modal-helpers.test to ts

DanielRosa74 committed:
Update tests and translations.json

DanielRosa74 committed:
fix translation positioning

DanielRosa74 committed:
add react-speech-recognition and replace recording logic

DanielRosa74 commented on client/src/templates/Challenges/components/speaking-modal.tsx:
```diff
@@ -0,0 +1,451 @@
+  const startRecording = async () => {
+    // Check speech recognition support first
+    const SupportedSpeechRecognition =
+      window.SpeechRecognition || window.webkitSpeechRecognition;
```
Changed to react-speech-recognition. I believe this renders it unnecessary.

DanielRosa74 commented on client/src/templates/Challenges/components/speaking-modal.tsx:
```diff
@@ -0,0 +1,451 @@
+    }
+  };
+
+  const stopRecording = () => {
```
Done.

DanielRosa74 commented on client/src/templates/Challenges/components/multiple-choice-questions.tsx:
```diff
@@ -57,29 +81,63 @@ function MultipleChoiceQuestions({
+                          marginBottom: '2px'
+                        }}
+                      >
+                        <button
```
UI/Button added.

DanielRosa74 commented on client/src/templates/Challenges/components/speaking-modal-helpers.test.js:
Changed to TypeScript. Hard-coded strings are now in translations.json and tests updated to contemplate that.

DanielRosa74 commented on client/src/templates/Challenges/components/speaking-modal.tsx:
```diff
@@ -0,0 +1,451 @@
+                      <span
+                        key={index}
+                        style={{
+                          color: item.isCorrect ? 'green' : 'red',
```
This was the only implementation I haven't touched yet so I could have a more detailed chat with you as to the best way of implementing sr-only to the code.

huyenltnguyen committed:
fix: modal size

huyenltnguyen committed:
fix: modal styling and a11y [skip ci]

huyenltnguyen reviewed (commented):
I just pushed some changes to the PR. It's mostly styling and a11y improvements:
- Changed the modal size to `large` so the modal is a little larger.
  - Btw, I forgot to share the link to the UI library documentation. In case you want to know what we components currently have / what variants a component has, you can go to https://opensource.freecodecamp.org/ui/.
- Minor CSS cleanup
- Replaced the label and read-only input elements in the modal with a `p` since the sentence is a static content and not a form control
- Turned the feedback `div` to a live region so the dynamic feedback is announced to screen reader users
- Refactored the feedback rendering

---

I'm still going through the PR so please bear with me 😄 

Here are a couple of things I noticed that I think are worth considering:
- For the record action, maybe we allow campers to stop recording and try again (we do have a 30-second timeout in the code but that's not visible to the users, and maybe they don't want to wait that long).
- For unsupported browsers, I think we should hide the speaking button because it's not super user-friendly: campers click the Speaking button → press Play → press Record → then discover their browser isn’t supported.
  - Instead, we could display a banner if the browser isn't supported.
    - We can use the `Callout` component from freecodecamp/ui here
  - We could also split the listening and speaking functionality into two:
    - Show one button for listening, and one for speaking
    - If the browser doesn't have speech recognition supported, hide the speaking button so at least campers can still practice listening
- On the challenge page, would it be cleaner if we use an icon button instead of a button with the text "Speaking"?
  - It's just that I think "Practice Speaking" is a little clearer than "Speaking". But that's too much text and could distract from the MCQ answers.
  - We could use the microphone icon in this case. And if we go forward with the feature split, then the headphone icon for the listening button.
  - Btw, we use Font Awesome icons, so you can find the available icons here: https://fontawesome.com/v6/icons#packs

---

I may have more thoughts as I get further into the review 😄 

Don’t feel obligated to make the changes, though. I'm happy to help implement them.

huyenltnguyen committed:
refactor: make compareTexts return status instead of message strings

huyenltnguyen committed:
fix(a11y): display utterance as a full sentence for screen readers

huyenltnguyen committed:
refactor: remove audio status from feedback, handle audio cleanup

huyenltnguyen committed:
fix: speaking-model-helpers tests

huyenltnguyen committed:
refactor: improve recording cleanup

huyenltnguyen committed:
chore: use translation strings [skip ci]

huyenltnguyen committed:
fix: timeout cleanup

huyenltnguyen committed:
feat: set speech language dynamically

huyenltnguyen committed:
fix(a11y): create two different sets of speaking buttons for better a11y

huyenltnguyen committed:
fix: use the punctuation mark of the original sentence

huyenltnguyen committed:
fix: allow stopping recording

huyenltnguyen committed:
refactor: add audioId to mcq answer object

huyenltnguyen committed:
fix: apply border styles correctly

huyenltnguyen committed:
Merge branch 'main' of https://github.com/freeCodeCamp/freeCodeCamp into pr/61906

huyenltnguyen committed:
fix: pin deps versions

huyenltnguyen committed:
fix: hide the record button if browser is not supported

huyenltnguyen committed:
feat: add icon to the buttons

huyenltnguyen committed:
fix: merge the two button sets back into one

huyenltnguyen committed:
test: add e2e tests for mcq

huyenltnguyen committed:
fix: update snapshot

huyenltnguyen committed:
refactor: minor cleanup

huyenltnguyen committed:
Merge branch 'main' of https://github.com/freeCodeCamp/freeCodeCamp into pr/61906

huyenltnguyen committed:
fix: import vitest testing functions

huyenltnguyen committed:
fix: lock file

ojeytonwilliams reviewed (dismissed):
This is looking much better, thanks @DanielRosa74 and @huyenltnguyen 

I didn't find much on this pass, but I did find one more bug:

ojeytonwilliams commented:
By the way: do you want to release with just that one lesson? Or do you want to revert that and do a bunch at once?

huyenltnguyen committed:
Apply suggestions from code review

Co-authored-by: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

huyenltnguyen committed:
refactor: convert render functions to components

huyenltnguyen committed:
refactor: remove changes to challenge file

huyenltnguyen committed:
refactor: remove showSpeakingButton

huyenltnguyen committed:
refactor: rename audioid to audio-id

huyenltnguyen committed:
test: skip speaking modal test suite

DanielRosa74 committed:
Merge branch 'main' into speaking-button

DanielRosa74 committed:
address speaker helper feedback changes

DanielRosa74 committed:
test: update challenge parser snapshot

ojeytonwilliams committed:
feat: re-enable speaking lessons

ojeytonwilliams committed:
DEBUG: add audio-id to challenge

ojeytonwilliams committed:
feat: show misplaced words

Also refactored a bit. alignWords returns the more generic { expected: string, actual: string } and the component decides how to render it

ojeytonwilliams committed:
fix: add audioId to schema so Gatsby can build

ojeytonwilliams committed:
feat: more consistent, better tested comparison

ojeytonwilliams committed:
refactor: fix types

ojeytonwilliams committed:
Merge branch 'main' into speaking-button

ojeytonwilliams committed:
fix: handle missing leading words correctly

ojeytonwilliams committed:
feat: include trailing extra words

ojeytonwilliams committed:
fix: allow showSpeakingButton

huyenltnguyen commented on e2e/multiple-choice-question-challenge.spec.ts:
```diff
@@ -0,0 +1,118 @@
+  '/learn/full-stack-developer/lecture-what-is-css/what-is-the-basic-anatomy-of-a-css-rule';
+
+// TODO: Enable this test suite when we have added audio to the challenge
+test.describe.skip(
```
The test was skipped back when we removed `audioId` from the challenge file. Now that we're adding it back, we should enable the test.

Ref: https://github.com/freeCodeCamp/freeCodeCamp/pull/61906#issuecomment-3313787372

```suggestion
test.describe(
```

huyenltnguyen commented on client/src/templates/Challenges/components/multiple-choice-questions.tsx:
```diff
@@ -12,6 +18,9 @@ type MultipleChoiceQuestionsProps = {
   handleOptionChange: (questionIndex: number, answerIndex: number) => void;
   submittedMcqAnswers: (number | null)[];
   showFeedback: boolean;
+
```
```suggestion
```

DanielRosa74 committed:
Apply suggestions from code review

Co-authored-by: Huyen Nguyen <25715018+huyenltnguyen@users.noreply.github.com>

huyenltnguyen committed:
fix: formatting

ojeytonwilliams reviewed (approved):
Then I shall pick up the approval baton
